### PR TITLE
Rabi with probabilities

### PR DESCRIPTION
**Pull request by Edoardo-Pedicillo**
As the title says.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.
- [ ] Compatibility with Qibo modules (Please edit this section if the current pull request is not compatible with the following branches).
    - [ ] Qibo: `master`
    - [ ] Qibolab: `main`
    - [ ] Qibolab_platforms_qrc: `main`

### CODE REVIEW

File: src/qibocal/update.py [hunk 1]
```diff
@@ -44,23 +44,24 @@ def readout_attenuation(att: int, platform: Platform, qubit: QubitId):
...
 
-def drive_amplitude(amp: float, platform: Platform, qubit: QubitId):
+def drive_amplitude(amp: Union[float, tuple], platform: Platform, qubit: QubitId):
     """Update drive frequency value in platform for specific qubit."""
+    if isinstance(amp, tuple):
+        amp = amp[0]
     platform.qubits[qubit].native_gates.RX.amplitude = float(amp)
 
 
 def drive_duration(duration: int, platform: Platform, qubit: QubitId):
```
Comment (GabrielePalazzo) [inline src/qibocal/update.py:61]:
```suggestion
def drive_duration(duration: Union[int, tuple], platform: Platform, qubit: QubitId):
```
I think we should add `tuple` if we support tuple[int] as input type.

**Review by GabrielePalazzo — COMMENTED** [review]
Could you also add the runcards used for running these routines?

File: src/qibocal/protocols/characterization/rabi/amplitude.py [hunk 2]
```diff
@@ -36,16 +37,21 @@ class RabiAmplitudeParameters(Parameters):
 class RabiAmplitudeResults(Results):
     """RabiAmplitude outputs."""
 
-    amplitude: dict[QubitId, float] = field(metadata=dict(update="drive_amplitude"))
+    amplitude: dict[QubitId, tuple[float, Optional[float]]] = field(
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/rabi/amplitude.py:40]: I just spotted an error that was already in main.
In all results attribute we can remove the `metadata...update` from field given that now the update is performed separately.
Could you please just remove it from the routines that you edited?

File: src/qibocal/protocols/characterization/rabi/amplitude.py [hunk 3]
```diff
@@ -111,21 +117,21 @@ def _acquisition(
...
-            averaging_mode=AveragingMode.CYCLIC,
+            acquisition_type=AcquisitionType.DISCRIMINATION,
+            averaging_mode=AveragingMode.SINGLESHOT,
         ),
         sweeper,
     )
     for qubit in qubits:
         # average msr, phase, i and q over the number of shots defined in the runcard
-        result = results[ro_pulses[qubit].serial]
+        prob = results[qubit].probability(state=1)
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/rabi/amplitude.py:127]: Outdated comment

File: src/qibocal/protocols/characterization/rabi/amplitude_msr.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    sweeper = Sweeper(
+        Parameter.amplitude,
+        qd_pulse_amplitude_range,
+        [qd_pulses[qubit] for qubit in qubits],
+        type=SweeperType.FACTOR,
+    )
+
+    # create a DataUnits object to store the results,
+    # DataUnits stores by default MSR, phase, i, q
+    # additionally include qubit drive pulse amplitude
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/rabi/amplitude_msr.py:87]: Outdated comments

File: src/qibocal/protocols/characterization/rabi/amplitude_msr.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        ExecutionParameters(
+            nshots=params.nshots,
+            relaxation_time=params.relaxation_time,
+            acquisition_type=AcquisitionType.INTEGRATION,
+            averaging_mode=AveragingMode.CYCLIC,
+        ),
+        sweeper,
+    )
+    for qubit in qubits:
+        # average msr, phase, i and q over the number of shots defined in the runcard
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/rabi/amplitude_msr.py:102]: Outdated comments

File: src/qibocal/protocols/characterization/rabi/length.py [hunk 6]
```diff
@@ -36,16 +37,21 @@ class RabiLengthParameters(Parameters):
...
     """RabiLength outputs."""
 
-    length: dict[QubitId, int] = field(metadata=dict(update="drive_length"))
+    length: dict[QubitId, tuple[int, Optional[float]]] = field(
+        metadata=dict(update="drive_length")
+    )
     """Pi pulse duration for each qubit."""
-    amplitude: dict[QubitId, float] = field(metadata=dict(update="drive_amplitude"))
+    amplitude: dict[QubitId, tuple[float, Optional[float]]] = field(
+        metadata=dict(update="drive_amplitude")
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/rabi/length.py:45]: Remove metadata (see comment above)

File: src/qibocal/protocols/characterization/rabi/length.py [hunk 7]
```diff
@@ -115,22 +121,22 @@ def _acquisition(
...
+            acquisition_type=AcquisitionType.DISCRIMINATION,
+            averaging_mode=AveragingMode.SINGLESHOT,
         ),
         sweeper,
     )
 
     for qubit in qubits:
-        # average msr, phase, i and q over the number of shots defined in the runcard
-        result = results[ro_pulses[qubit].serial]
+        # average prob, phase, i and q over the number of shots defined in the runcard
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/rabi/length.py:131]: Outdated comment

File: src/qibocal/protocols/characterization/rabi/length_msr.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    sweeper = Sweeper(
+        Parameter.duration,
+        qd_pulse_duration_range,
+        [qd_pulses[qubit] for qubit in qubits],
+        type=SweeperType.ABSOLUTE,
+    )
+
+    # create a DataUnits object to store the results,
+    # DataUnits stores by default MSR, phase, i, q
+    # additionally include qubit drive pulse length
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/rabi/length_msr.py:91]: Oudated comment

File: src/qibocal/protocols/characterization/rabi/utils.py [hunk 9]
```diff
@@ -116,3 +116,82 @@ def plot(data, qubit, fit):
     figures.append(fig)
 
     return figures, fitting_report
+
+
+def plot_proba(data, qubit, fit):
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/rabi/utils.py:121]:
```suggestion
def plot_probabilities(data, qubit, fit):
```
This name should be better for readability

**Review by andrea-pasquale — COMMENTED** [review]
Thanks @Edoardo-Pedicillo.
Looks good to me, you can find below some comments (mainly documentation).
I still need to test on hardware.

**Review by andrea-pasquale — APPROVED** [review]
Thanks for implementing this @Edoardo-Pedicillo.
LGTM, tested on zh and it works.

**Review by GabrielePalazzo — APPROVED** [review]

**Pull request merged**